_posts/pl/2021-01-28-migracja-schematow-bazy-danych.md

```diff
@@ -0,0 +1,108 @@
+* Zalecane jest, aby każda zmiana była jak najmniejsza i najlepiej możliwa do odwrócenia. 
+  Przykładowo tworząc indeksy na istniejących tabelach, najlepiej rozbić ich tworzenie do osobnych wersji.
+  Jeśli nie zostaną one wykonane w osobnych migracjach schematu, wtedy narażamy się na ryzyko takie jak opisane niżej:
+  * w tej samej migracji schematu bazy danych tworzymy indeks A oraz indeks B,
+  * stworzenie indeksu A zajmuje 5 minut i przebiega poprawnie,
+  * stworzenie indeksu B zajmuje ponad 5 minut i powoduje błąd TimeoutException,
+  * oba indeksy zostają wycofane i indeks A musi być ponownie założony,
+  * w przypadku, gdyby tworzenie indeksów było rozdzielone na osobne migracje, wtedy nie będzie konieczności ponownego tworzenia indeksu A (i ponownego poświęcania 5 minut na ten cel).
+* Wykonywane zmiany powinny być przyrostowe, czyli zmiana dla danej wersji powinna być uruchomiona tylko raz.
+
+#### Przykładowe biblioteki
+* Liquibase
+  * [liquibase.org](https://www.liquibase.org/)
+* FlywayDB
+  * [flywaydb.org](https://flywaydb.org/)
+* MyBatis Migration
+  * [mybatis.org/migrations](https://mybatis.org/migrations)
+* w przypadku bazy MongoDB
+  * Mongock
+    * [github.com/cloudyrock/mongock](https://github.com/cloudyrock/mongock)
+    * [mongock.io](https://www.mongock.io)
+  * Migrate-mongo
+    * [npmjs.com/package/migrate-mongo](https://www.npmjs.com/package/migrate-mongo)
+    * [github.com/seppevs/migrate-mongo](https://github.com/seppevs/migrate-mongo)
+  * liquibase-mongodb
+    * [github.com/liquibase/liquibase-mongodb](https://github.com/liquibase/liquibase-mongodb)
+
+w powyższych bibliotekach tworzenie i uruchamianie przyrostowe zmian jest wbudowane.
+
+### Zmiany powinny być kompatybilne wstecz
+Jest to szczególnie ważne, jeśli chcemy bezprzerwowo aktualizować naszą bazę danych, w takim wypadku aktualizacja bazy danych nie powinna spowodować, że starsza wersja aplikacji przestanie działać.
+
+Przykładowo nowa kolumna powinna mieć domyślną wartość lub przyjmować null-e.
+
+Zmiana nazwy kolumny lub jej usunięcie powinno być rozbite na kilka etapów, tak aby jej prawdziwe usunięcie było wykonane nie w docelowej wersji, tylko np. w następnej iteracji, gdy będziemy pewni, że żadna aplikacja z niej nie korzysta.
+
+## Aplikowanie zmian
+Aplikowania zmian wykonanych w ramach ewolucyjnej bazy danych jest już zależne od konkretnego przypadku.
+
+Jeśli baza danych jest ściśle związana jedną z aplikacją, możemy ją uruchamiać bezpośrednio z kodu [🔗⁵](https://andrewlock.net/deploying-asp-net-core-applications-to-kubernetes-part-7-running-database-migrations/#running-migrations-on-application-startup) .
+
+W przypadku gdy aplikacja jest rozproszona i nie chcemy blokować wszystkich instancji aplikacji na czas migracji schematu lub gdy kilka różnych aplikacji korzysta z tej bazy danych, możemy uruchamiać migrację niezależnie od aplikacji.
```

Comment:
Trochę nie wiadomo z czego wynika wypunktowanie więc dałbym chyba na końcu tego zdania: "W tym przypadku mamy następujące możliwości:" (lub coś podobnego). I jeszcze zmieniłbym to wypunktowanie na unordered.

Reply:
done